Simplenote/AuthenticationValidator.swift

```diff
@@ -0,0 +1,132 @@
+import Foundation
+
+
+// MARK: - AuthenticationValidator
+//
+struct AuthenticationValidator {
+
+    /// We're enhancing password strength requirements. But legacy accounts must be allowed in. That's why we have two sets of rules!
+    ///
+    let style: Style
+
+    /// Minimum Password Length: Login
+    ///
+    private let loginPasswordLength  = UInt(4)
+
+    /// Minimum Password Length: SignUp
+    ///
+    private let signupPasswordLength = UInt(8)
+
+    /// Defines the minimum allowed password length
+    ///
+    private var minimumPasswordLength: UInt {
+        return (style == .login) ? loginPasswordLength : signupPasswordLength
+    }
+
+
+    /// Returns the Validation Result for a given Username
+    ///
+    func performUsernameValidation(username: String) -> Result {
+        let predicate = NSPredicate.predicateForEmailValidation()
+        return predicate.evaluate(with: username) ? .success : .emailInvalid
+    }
+
+    /// Returns the Validation Result for a given Password (with its associated Username)
+    ///
+    func performPasswordValidation(username: String, password: String) -> Result {
+        guard password.count >= minimumPasswordLength else {
+            return .passwordTooShort(length: minimumPasswordLength)
+        }
+
+        guard style == .signup else {
+            return .success
+        }
+
+        guard password != username else {
+            return .passwordMatchesUsername
+        }
+
+        guard !password.contains(String.newline), !password.contains(String.tab) else {
+            return .passwordContainsInvalidCharacter
+        }
+
+        return .success
+    }
+}
+
+
+// MARK: - Nested Types
+//
+extension AuthenticationValidator {
+
+    enum Style {
+        case login
+        case signup
+    }
+
+    enum Result {
+        case success
+        case emailInvalid
+        case passwordMatchesUsername
+        case passwordTooShort(length: UInt)
+        case passwordContainsInvalidCharacter
+    }
+}
+
+
+// MARK: - Validation Results: String Conversion
+//
+extension AuthenticationValidator.Result: CustomStringConvertible {
+
+    var description: String {
+        switch self {
+        case .success:
+            // Not really needed. But for convenience reasons, it's super if this property isn't optional.
+            return String()
+
```

Comment:
I ended up in the exact same corner 😄 !!! Writing them down for the record:

1. It made sense for `performValidation` not to throw (because a validation error would not be... an exception per se).
2. Using an optional value in `performValidation()` or `description` would have complicated the code.

So all in all, the easiest thing to do was: Valid / Invalid are enum cases, and if the caller uses (or not) the `.description`, it's not the validator's problem.

Thank you for bringing this up!!